src/Common/src/TypeSystem/Common/TypeSystemHelpers.cs

```diff
@@ -36,18 +36,24 @@ static public PointerType MakePointerType(this TypeDesc type)
 
         static public DefType GetClosestDefType(this TypeDesc type)
         {
-            if (type is DefType)
-                return (DefType)type;
-            else
-                return type.BaseType;
+            // The implementation for this should mirror GetClosestMetadataType
+            throw new NotImplementedException();
         }
 
         static public MetadataType GetClosestMetadataType(this TypeDesc type)
         {
-            if (type is MetadataType)
-                return (MetadataType)type;
-            else
-                return type.BaseType.GetClosestMetadataType();
+            if (type.IsSzArray)
+            {
+                MetadataType arrayType = type.Context.SystemModule.GetType("System", "Array`1");
```

Comment:
I was playing around with this a bit. It feels like moving the `System.Array'1` reference out of here is just an abstraction for the sake of abstraction. There's little value in that.

The thing we would want to abstract out is `GetClosestMetadataType`. The method is exclusively used as a precondition to calling `EnumAllVirtualSlots`, `ResolveInterfaceMethodToVirtualMethodOnType`, and `FindVirtualFunctionTargetMethodOnObjectType`. It feels like these virtual resolution APIs put the onus of dealing with array types on the caller, instead of treating arrays specially themselves.

@davidwrighton said he doesn't like the idea of `ArrayType.GetMethods()` returning the explicitly implemented interface methods, but for this incarnation of the type system, it feels pretty natural - `GetMethods` already contains the `Address` method and friends. I could image having `ArrayType` call into type system context asking for its interface implementation methods. The MethodDescs (ideally `ILStubMethod` descendants) returned by context would call the ArrayType their owning type, and the IL could be backed by anything (methods on Array&lt;T&gt;, or even generic methods on non-generic types).

Extending `EnumAllVirtualSlots` and `FindVirtualFunctionTargetMethodOnObjectType` to accept ArrayType on top of MetadataType seems straightforward. Only `ResolveInterfaceMethodToVirtualMethodOnType` would be more tricky, but it can be easily solved with a new method on `ArrayType` (placed in `ArrayType.MethodImpls.cs`).

Thoughts? We could also postpone discussion on this by just merging this as-is and filing an issue to clean it up later...